src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php

```diff
@@ -32,10 +35,10 @@ protected function configure()
     {
         $this
             ->setName('security:encode-password')
-            ->setDescription('Encode a password.')
+            ->setDescription('Encodes a password.')
             ->addArgument('password', InputArgument::OPTIONAL, 'Enter a password')
-            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.')
-            ->addArgument('salt', InputArgument::OPTIONAL, 'Enter the salt you want to use to encode your password.')
+            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.', 'Symfony\Component\Security\Core\User\User')
+            ->addOption('empty-salt', null, InputOption::VALUE_NONE, 'Do not generate a salt or let the encoder generate one.')
```

Comment:
Thank you @fabpot 

@javiereguiluz "never assume anything on behalf of the developer" is a guideline. It's a **great** guideline. But it's not dogma. I don't think the meaning of the guideline is "we will let you shoot yourself in the foot if you want because theoretically you might _need_ to shoot yourself in the foot", it's more like "if we assume too much we risk making your job harder not easier" - this is not one of those instances. The only legitimate use-case I can think of for providing a known salt to the encoder is to verify its output (i.e. to test that it works), but that's not the job of this command.

I don't think your use-cases are legitimate, sorry. Frankly I'm not even convinced that _my_ use case is legitimate. They're theoretically plausible, but the solution to them isn't to make this command generate the salt, it's to make a custom PasswordEncoderInterface implementation that generates conforming salts in the same way that the BCrypt one does.

Basically this whole argument has come about because the design of that component is flawed - it's a leaky abstraction. Encoders right now demand that we have special knowledge of the length and composition of the salt, which is wrong. We don't know and can't know if the encoder accepts raw binary salts or base64 or (god help us) only hexits; we don't know how long the salt needs to be. This is wrong. Until that's fixed, I think we should do the best we can, which is to generate a reasonably long base64 salt for everything except BCrypt. Accepting a salt override (or even an empty-salt) is a mistake - it shouldn't be any of our business (in the context of this command), and it's a burden and a problem for users, not a benefit.